cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -318,6 +319,32 @@ private void getScheduleStatus(HttpResponder responder, String appId, String nam
     }
   }
 
+  /**
+   * Stops the particular run of the Workflow or MapReduce program.
+   */
+  @POST
+  @Path("/apps/{app-id}/{type}/{id}/runs/{run-id}/stop")
+  public void performRunLevelStop(HttpRequest request, HttpResponder responder,
+                                  @PathParam("namespace-id") String namespaceId,
+                                  @PathParam("app-id") String appId,
+                                  @PathParam("type") String type,
+                                  @PathParam("id") String id,
+                                  @PathParam("run-id") String runId) {
+    try {
+      ProgramType programType = ProgramType.valueOfCategoryName(type);
+      if (!isRunLevelActionAllowed(programType)) {
+        responder.sendString(HttpResponseStatus.FORBIDDEN,
```

Comment:
@chtyim @sagarkapare sorry, I missed this earlier. There are exceptions like ``CannotBeDeletedException``, ``CannotBeCreatedException``, etc in the codebase already. We respond with ``FORBIDDEN`` for these exceptions too.

I think it is ok to add ``ForbiddenException``, but lets also then make ``CannotBeDeletedException`` and ``CannotBeCreatedException`` subclasses of ``ForbiddenException``.

Also, I'm not sure, but should we rename it to ``CannotPerformOperationException`` to be consistent with existing names?